File: src/programmers/heap/Programmers1.java
package programmers.heap;

import java.util.PriorityQueue;

public class Programmers1 {
    private int solution(int[] scoville, int K) {
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>();

        for (Integer scov : scoville) {
            priorityQueue.offer(scov);
        }

        int answer = 0;
        while (!priorityQueue.isEmpty()) {
            int minimum = priorityQueue.poll();
            if(minimum >= K) {
                priorityQueue.offer(minimum);
                break;
            }
            if (priorityQueue.isEmpty()) {
                answer = -1;
                break;
            }
            int secondMin = priorityQueue.poll();

            int mixValue = minimum + (secondMin * 2);
            answer++;
            priorityQueue.offer(mixValue);
        }

        return answer;
    }

    public static void main(String[] args) {
        System.out.println(new Programmers1().solution(new int[]{1, 2, 3, 9, 10, 12}, 7));
    }
}

File: src/programmers/queue/Programmers1.kt
package programmers.queue

import java.util.*

class Programmers1 {
    fun solution(progresses: IntArray, speeds: IntArray): IntArray {
        val jobPeriod = LinkedList<Int>()
        for (i in 0 until progresses.size) {
            val remains = 100 - progresses[i]
            var period = remains / speeds[i]

            if (remains % speeds[i] != 0) period++

            jobPeriod.offer(period)
        }

        val answer = mutableListOf<Int>()
        var criteria = jobPeriod.poll()
        var cnt = 1
        while (!jobPeriod.isEmpty()) {
            val p = jobPeriod.poll()

            if (p <= criteria) {
                cnt++
            } else {
                answer.add(cnt)
                criteria = p
                cnt = 1
            }
        }
        answer.add(cnt)

        return answer.toIntArray()
    }
}

fun main(args: Array<String>) {
    Programmers1().solution(intArrayOf(93, 30, 55), intArrayOf(1, 30, 5)).forEach { println(it) }
}